packages/itwin/grouping-mapping-widget/src/components/Mappings/hooks/useFetchExtractionStates.ts

```diff
@@ -0,0 +1,76 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+import { useExtractionClient } from "../../context/ExtractionClientContext";
+import { ExtractorState } from "@itwin/insights-client";
+import { useMutation, useQuery } from "@tanstack/react-query";
+import type { GroupingMappingApiConfig } from "../../context/GroupingApiConfigContext";
+import type { ExtractionRequestMapping, ExtractionRunRequest, Mapping } from "@itwin/insights-client";
+import { useState } from "react";
+import { ExtractionStates } from "../Extraction/ExtractionStatus";
+import { STATUS_CHECK_INTERVAL } from "../../Constants";
+
+export const useFetchExtractionStates = ({
+  iModelId,
+  getAccessToken,
+}: GroupingMappingApiConfig) => {
+
+  const extractionClient = useExtractionClient();
+  const [isJobStarted, setIsJobStarted] = useState<boolean>(false);
+  const [isJobDone, setIsJobDone] = useState<boolean>(false);
+
+  const { mutateAsync: runExtraction } = useMutation({
+    mutationKey: ["runExtraction"],
+    mutationFn: async (mappings: Mapping[]) => {
+      const accessToken = await getAccessToken();
+      const mappingIds: ExtractionRequestMapping[] = mappings.length > 0 ? mappings.map((mapping) => { return { id: mapping.id }; }) : [];
+      const extractionRequest: ExtractionRunRequest | undefined = {
+        mappings: mappingIds,
+      };
+      await extractionClient.runExtraction(accessToken, iModelId, extractionRequest);
+    },
+  });
+
+  const { data: extractionStateData, isError: isExtractionStateError, isLoading: isExtractionStatusLoading } = useQuery({
+    queryKey: ["extractionState"],
+    staleTime: Infinity,
+    placeholderData: undefined,
+    queryFn: async () => {
+      const accessToken = await getAccessToken();
+      const extractionHistoryIterator = extractionClient.getExtractionHistoryIterator(accessToken, iModelId, 1);
```

Comment:
Querying extraction history on every state check if very inefficient. The state checks themselves cause a lot of network traffic, since it is polling for state every couple of seconds. Adding the extraction history query will double the amount of network traffic.

Reply:
I have worked around and found a way to cache the jobId accordingly to the selected mapping(s) every time users run extraction so when it comes to querying status, it refers to that cached jobId without going to the route of retrieving history.

Reply:
What if user refreshes the web page? Then all of the state is lost. So we still need to have the extraction history query somewhere. Completely removing it is not the right way.

Reply:
Personally, I think having jobID mapped with each mapping then using `getExtractionStatus` to retrieve status for that specific mapping will be sufficient since users can start extraction on mapping(s) at different time. Extraction history will potential return the latest job(s) but it might cause running mappings to reset their status as soon as another job is started. That is also why I stored a hash map in context with keys are mapping IDs and values are their assigned jobIDs.

Reply:
But I think it's a good idea to think of a way to keep/store extraction status such as passing down a list of query results when `Mapping` mounts so that we won't lose status when users refresh the whole page.